File: litho-core/src/main/java/com/facebook/litho/EquivalenceUtils.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.litho;

import android.util.SparseArray;
import androidx.annotation.Nullable;
import com.facebook.litho.drawable.ComparableColorDrawable;
import com.facebook.rendercore.primitives.Equivalence;
import java.lang.reflect.Field;
import java.util.Arrays;
import java.util.Collection;
import java.util.Iterator;

public class EquivalenceUtils {

  /** Checks if objects are equal. */
  public static boolean equals(@Nullable Object a, @Nullable Object b) {
    if (a == b) {
      return true;
    }

    if (a == null || b == null) {
      return false;
    }

    return a.equals(b);
  }

  /** Checks if {@link SparseArray} objects are equal. */
  public static boolean equals(@Nullable SparseArray<?> a, @Nullable SparseArray<?> b) {
    if (a == b) {
      return true;
    }

    if (a == null || b == null) {
      return false;
    }

    if (a.size() != b.size()) {
      return false;
    }

    int size = a.size();

    for (int i = 0; i < size; i++) {
      if (a.keyAt(i) != b.keyAt(i)) {
        return false;
      }

      if (!equals(a.valueAt(i), b.valueAt(i))) {
        return false;
      }
    }

    return true;
  }

  /** Checks if {@link Equivalence} objects are equivalent. */
  public static <T extends Equivalence<T>> boolean isEquivalentTo(@Nullable T a, @Nullable T b) {
    if (a == b) {
      return true;
    }

    if (a == null || b == null) {
      return false;
    }

    return a.isEquivalentTo(b);
  }

  /**
   * Checks if to objects are equivalent if they implement {@link Equivalence} or compares the
   * objects field by field.
   */
  public static <T> boolean isEqualOrEquivalentTo(@Nullable T a, @Nullable T b) {
    if (a instanceof Equivalence && b instanceof Equivalence) {
      return isEquivalentTo((Equivalence) a, (Equivalence) b);
    } else {
      return hasEquivalentFields(a, b, false);
    }
  }

  /** Compare all private final fields in an object */
  @SuppressWarnings("unchecked")
  public static boolean hasEquivalentFields(Object a, Object b) {
    return hasEquivalentFields(a, b, true);
  }

  /** Compare all private final fields in an object */
  @SuppressWarnings("unchecked")
  public static boolean hasEquivalentFields(Object a, Object b, boolean shouldCompareCommonProps) {
    if (a == b) {
      return true;
    }
    if (a.getClass() != b.getClass()) {
      return false;
    }

    for (Field field : a.getClass().getDeclaredFields()) {
      final Object val1;
      final Object val2;
      try {
        final boolean wasAccessible = field.isAccessible();
        if (!wasAccessible) {
          field.setAccessible(true);
        }
        val1 = field.get(a);
        val2 = field.get(b);
        if (!wasAccessible) {
          field.setAccessible(false);
        }
      } catch (IllegalAccessException e) {
        throw new IllegalStateException("Unable to get fields by reflection.", e);
      }

      if (!EquivalenceUtils.areObjectsEquivalent(val1, val2, shouldCompareCommonProps)) {
        return false;
      }
    }

    return true;
  }

  /** Return true if two objects are equivalent. */
  @SuppressWarnings("unchecked")
  public static boolean areObjectsEquivalent(
      @Nullable Object val1, @Nullable Object val2, boolean shouldCompareCommonProps) {
    if (val1 == val2) {
      return true;
    }
    if (val1 == null || val2 == null || val1.getClass() != val2.getClass()) {
      return false;
    }

    if (val1 instanceof Float) {
      return Float.compare((Float) val1, (Float) val2) == 0;
    } else if (val1 instanceof Double) {
      return Double.compare((Double) val1, (Double) val2) == 0;
    } else if (val1 instanceof Component) {
      return ((Component) val1).isEquivalentTo((Component) val2, shouldCompareCommonProps);
    } else if (val1 instanceof Equivalence) {
      return ((Equivalence) val1).isEquivalentTo(val2);
    } else if (val1.getClass().isArray()) {
      // TODO(T69494307): replace this with areArraysEquivalent(Object, Object).
      return ComponentUtils.areArraysEquals(val1.getClass(), val1, val2);
    } else if (val1 instanceof Collection) {
      // TODO(T69494307):
      // replace this with areCollectionsEquivalent(Collection, Collection).
      return ComponentUtils.areCollectionsEquals(
          val1.getClass().getGenericSuperclass(), (Collection) val1, (Collection) val2);
    } else if (val1 instanceof ComparableColorDrawable) {
      return ((ComparableColorDrawable) val1).isEquivalentTo((ComparableColorDrawable) val2);
    } else if (val1 instanceof EventHandler) {
      return ((EventHandler) val1).isEquivalentTo((EventHandler) val2);
    }
    return val1.equals(val2);
  }

  /**
   * TODO(T69494307): Don't delete this method, it's going to replace {@code
   * ComponentUtils.areArraysEquals(Class<?>, Object, Object)}
   */
  private static boolean areArraysEquivalent(
      Object val1, Object val2, boolean shouldCompareCommonProps) {
    if (val1 instanceof byte[]) {
      return Arrays.equals((byte[]) val1, (byte[]) val2);
    } else if (val1 instanceof short[]) {
      return Arrays.equals((short[]) val1, (short[]) val2);
    } else if (val1 instanceof char[]) {
      return Arrays.equals((char[]) val1, (char[]) val2);
    } else if (val1 instanceof int[]) {
      return Arrays.equals((int[]) val1, (int[]) val2);
    } else if (val1 instanceof long[]) {
      return Arrays.equals((long[]) val1, (long[]) val2);
    } else if (val1 instanceof float[]) {
      return Arrays.equals((float[]) val1, (float[]) val2);
    } else if (val1 instanceof double[]) {
      return Arrays.equals((double[]) val1, (double[]) val2);
    } else if (val1 instanceof boolean[]) {
      return Arrays.equals((boolean[]) val1, (boolean[]) val2);
    } else {
      final Object[] array1 = (Object[]) val1;
      final Object[] array2 = (Object[]) val2;

      if (array1.length != array2.length) {
        return false;
      }

      for (int i = 0, size = array1.length; i < size; i++) {
        if (!areObjectsEquivalent(array1[i], array2[i], shouldCompareCommonProps)) {
          return false;
        }
      }
    }
    return true;
  }

  /**
   * TODO(T69494307): Don't delete this method, it's going to replace {@code
   * ComponentUtils.areCollectionsEquals(Type, Collection, Collection)}
   */
  private static boolean areCollectionsEquivalent(
      Collection collection1, Collection collection2, boolean shouldCompareCommonProps) {
    if (collection1.size() != collection2.size()) {
      return false;
    }

    final Iterator iterator1 = collection1.iterator();
    final Iterator iterator2 = collection2.iterator();

    while (iterator1.hasNext()) {
      final Object elem1 = iterator1.next();
      final Object elem2 = iterator2.next();
      if (!areObjectsEquivalent(elem1, elem2, shouldCompareCommonProps)) {
        return false;
      }
    }
    return true;
  }
}

File: litho-rendercore-primitives/src/main/java/com/facebook/rendercore/primitives/utils/EquivalenceUtils.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

@file:JvmName("EquivalenceUtils")

package com.facebook.rendercore.primitives.utils

import android.util.SparseArray
import com.facebook.rendercore.primitives.Equivalence

/** Checks if two objects are equal. */
fun equals(a: Any?, b: Any?): Boolean {
  return a == b
}

/** Checks if two [SparseArray] objects are equal. */
fun equals(a: SparseArray<*>?, b: SparseArray<*>?): Boolean {
  if (a === b) {
    return true
  }

  if (a == null || b == null) {
    return false
  }

  val size = a.size()
  if (size != b.size()) {
    return false
  }

  for (index in 0 until size) {
    val key: Int = a.keyAt(index)
    if (!equals(a.valueAt(index), b.get(key))) {
      return false
    }
  }

  return true
}

/** Checks if two [Equivalence] objects are equivalent. */
fun <T : Equivalence<T>> isEquivalentTo(a: T?, b: T?): Boolean {
  return if (a === b) {
    true
  } else if (a == null || b == null) {
    false
  } else {
    a.isEquivalentTo(b)
  }
}

/**
 * Checks if two objects are equivalent if they implement [Equivalence] or compares the objects
 * field by field.
 */
@Suppress("UNCHECKED_CAST")
fun <T> isEqualOrEquivalentTo(a: T?, b: T?): Boolean {
  return if (a === b) {
    true
  } else if (a == null || b == null) {
    false
  } else if (a is Equivalence<*> && b is Equivalence<*>) {
    isEquivalentTo(a as Equivalence<Any>, b as Equivalence<Any>)
  } else {
    hasEquivalentFields(a, b)
  }
}

/** Checks if all private final fields in two objects are equivalent. */
fun hasEquivalentFields(a: Any?, b: Any?): Boolean {
  if (a === b) {
    return true
  }

  if (a == null || b == null) {
    return false
  }

  if (a.javaClass != b.javaClass) {
    return false
  }

  for (field in a.javaClass.declaredFields) {
    val val1: Any?
    val val2: Any?
    try {
      val wasAccessible = field.isAccessible
      if (!wasAccessible) {
        field.isAccessible = true
      }
      val1 = field[a]
      val2 = field[b]
      if (!wasAccessible) {
        field.isAccessible = false
      }
    } catch (e: IllegalAccessException) {
      throw IllegalStateException("Unable to get fields by reflection.", e)
    }

    if (!areObjectsEquivalent(val1, val2)) {
      return false
    }
  }

  return true
}

/** Checks if two objects are equivalent. */
@Suppress("UNCHECKED_CAST")
fun areObjectsEquivalent(val1: Any?, val2: Any?): Boolean {
  if (val1 === val2) {
    return true
  }

  if (val1 == null || val2 == null || val1.javaClass != val2.javaClass) {
    return false
  }

  return when {
    val1 is Float -> val1.compareTo(val2 as Float) == 0
    val1 is Double -> val1.compareTo(val2 as Double) == 0
    val1 is Equivalence<*> -> (val1 as Equivalence<Any>).isEquivalentTo(val2)
    val1.javaClass.isArray -> areArraysEquivalent(val1, val2)
    val1 is List<*> && val1 is RandomAccess -> areRandomAccessListsEquivalent(val1, val2 as List<*>)
    val1 is Collection<*> -> areCollectionsEquivalent(val1, val2 as Collection<*>)
    else -> val1 == val2
  }
}

/** Checks if two arrays are equivalent. */
@Suppress("UNCHECKED_CAST")
private fun areArraysEquivalent(val1: Any, val2: Any): Boolean {
  return when (val1) {
    is ByteArray -> val1.contentEquals(val2 as ByteArray)
    is ShortArray -> val1.contentEquals(val2 as ShortArray)
    is CharArray -> val1.contentEquals(val2 as CharArray)
    is IntArray -> val1.contentEquals(val2 as IntArray)
    is LongArray -> val1.contentEquals(val2 as LongArray)
    is FloatArray -> val1.contentEquals(val2 as FloatArray)
    is DoubleArray -> val1.contentEquals(val2 as DoubleArray)
    is BooleanArray -> val1.contentEquals(val2 as BooleanArray)
    else -> {
      val array1 = val1 as Array<Any>
      val array2 = val2 as Array<Any>

      if (array1.size != array2.size) {
        return false
      }

      val size = array1.size
      for (i in 0 until size) {
        if (!areObjectsEquivalent(array1[i], array2[i])) {
          return false
        }
      }

      return true
    }
  }
}

/** Checks if two random access lists are equivalent. */
private fun areRandomAccessListsEquivalent(val1: List<*>, val2: List<*>): Boolean {
  if (val1.size != val2.size) {
    return false
  }

  for (index in 0 until val1.size) {
    if (!areObjectsEquivalent(val1[index], val2[index])) {
      return false
    }
  }

  return true
}

/** Checks if two collections are equivalent. */
private fun areCollectionsEquivalent(val1: Collection<*>, val2: Collection<*>): Boolean {
  if (val1.size != val2.size) {
    return false
  }

  val iterator1 = val1.iterator()
  val iterator2 = val2.iterator()

  while (iterator1.hasNext()) {
    val elem1 = iterator1.next()
    val elem2 = iterator2.next()
    if (!areObjectsEquivalent(elem1, elem2)) {
      return false
    }
  }

  return true
}
